docs/content/framework/interfaces/DerivativeMaterialInterface.md

```diff
@@ -0,0 +1,120 @@
+# DerivativeMaterialInterface
+
+In model development, in particular using the finite element method, it is often convenient to encapsulate expression values _together with their derivatives_. The derivatives are often needed to compute Jacobian matrix entries.
+
+In MOOSE we use material properties to provide values for expressions that can be directly computed from variables in the solve. To store the derivatives of these material properties with respect to the variables they depend on we also use material properties. We use a standardized naming scheme to name the material property derivatives. The derivative of a material property `F` with respect to a variable `c` would be named `dF/dc`.
+
+To enforce this naming scheme we provide the ```DerivativeMaterialInterface```, a _veneer_ template that provides methods to _declare_ and _get_ material properties that are derivatives of other material properties.
+
+The ```DerivativeMaterialInterface``` is utilized by inheriting from it and supplying the original parent class as a template argument.
+
+# Examples
+
+The  can be used from Materials (to declare and get material property derivatives) or from Kernels (to get material property derivatives).
+
+## Use in a Material class
+
+### Header
+```c++
+#include "DerivativeMaterialInterface.h"
+#include "Material.h"
+
+//...
+
+// we template the interface on 'Material'
+class MyMaterial : public DerivativeMaterialInterface<Material>
+{
+
+//...
+
+private:
+  MaterialProperty<Real> & F;
+  MaterialProperty<Real> & dFdc;
+  MaterialProperty<Real> & d2Fdc2;
+};
+```
+
+Note that it is possible to template the interface on arbitrary classes, including classes derived from ```Material``` or ```Kernel```.
+
+### Implementation
+```c++
+#include "MyMaterial.h"
+
+[..]
+
+MyMaterial::MyMaterial(const InputParameters & parameters) :
+    DerivativeMaterialInterface<Material>(parameters),
+    // get the c variable value, number, and name
+    _c(coupledValue("c")),
+    _c_var(coupled("c")),
+    _c_name(getVar("c", 0)->name()),
+    // declare material property and first and second derivative w.r.t. c
+    F(declareProperty<Real>("F")),
+    dFdc(declarePropertyDerivative<Real>("F", _c_name))
+    d2Fdc2(declarePropertyDerivative<Real>("F", _c_name, _c_name))
+{
+}
+```
+
+Material property derivatives do _not_ need to be declared or computed in the same Material as their corresponding undifferentiated properties (though oftentimes it is a natural choice that provides encapsulation).
+
+## Use in a Kernel class
+
+### Header
+```c++
+#include "DerivativeMaterialInterface.h"
+#include "Kernel.h"
+
+[..]
```

Comment:
Well I think you guys are talking about different things. The "outdated diff" stuff seems to have disappeared which is unfortunate. The change is there if you click on the commit. This is a step back as far as I'm concerned.